### PR TITLE
ProtoColumn

### PR DESCRIPTION
**Pull request by acostama**
- we can serialize wire generated protobuf classes using wires adapters

### CODE REVIEW

**Comment by ryanhall07** [comment]
@hamdanjaveed you'lll be interested in using this

File: misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumn.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package misk.hibernate
+
+/**
+ * Put this on a column field for a class generated by Wire to get Hibernate to persist
+ * with protobuf serialization.
+ *
+ * It is an error to put this annotation on a mutable field. We don't defensively copy these when
```
Comment (ryanhall07) [inline misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumn.kt]: I don't follow this comment. it's obviously a copy of the data that is in the actual database

Comment (swankjesse) [reply misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumn.kt]: Protos are typically immutable. For JSON this comment was more appropriate where you could assign a thing and then mutate it

File: misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumnType.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.Properties
+
+internal class ProtoColumnType<T: Message<T,*>> : UserType, ParameterizedType {
+  private lateinit var protoAdapter: ProtoAdapter<T>
+
+
+  @Suppress("UNCHECKED_CAST")
+  override fun setParameterValues(properties: Properties) {
+    val clazz = properties.getField("protoColumnField")!!.type
+    val adapterField = clazz.getDeclaredField("ADAPTER")
```
Comment (swankjesse) [inline misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumnType.kt]: https://square.github.io/wire/2.x/wire-runtime/com/squareup/wire/ProtoAdapter.html#get-java.lang.Class-

File: misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumnType.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  @Suppress("UNCHECKED_CAST")
+  override fun nullSafeSet(
+    st: PreparedStatement,
+    value: Any?,
+    index: Int,
+    session: SharedSessionContractImplementor?
+  ) {
+    if (value == null ) {
+      st.setNull(index, Types.CHAR)
```
Comment (swankjesse) [inline misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumnType.kt]: CHAR ?

File: misk-hibernate/src/test/kotlin/misk/hibernate/ProtoColumnTest.kt [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+  @Test
+  fun basicPath() {
+    transacter.transaction { session ->
+      session.save(DbAvengersMovie(
+          name = "CivilWar",
+          hero = SuperHero.Builder()
+              .civilian_name("Tony Stark")
+              .super_hero_name("IronMan")
+              .powers(listOf(SuperPower("suit", 10), SuperPower("humor", 2)))
```
Comment (swankjesse) [inline misk-hibernate/src/test/kotlin/misk/hibernate/ProtoColumnTest.kt]: Nice

File: misk-hibernate/src/test/kotlin/misk/hibernate/ProtoColumnTest.kt [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+          .super_hero_name("IronMan")
+          .powers(listOf(SuperPower("suit", 10), SuperPower("humor", 2)))
+          .build()))
+    }
+    transacter.transaction { session ->
+      val movie = queryFactory.newQuery(DbAvengersMovieQuery::class)
+          .name("IronMan")
+          .nameAndHero(session)[0]
+      assertThat(movie.name).isEqualTo("IronMan")
+      assertThat(movie.hero).isEqualTo(SuperHero.Builder()
```
Comment (swankjesse) [inline misk-hibernate/src/test/kotlin/misk/hibernate/ProtoColumnTest.kt]: assert anotherHero is null?

File: misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumn.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package misk.hibernate
+
+/**
+ * Put this on a column field for a class generated by Wire to get Hibernate to persist
+ * with protobuf serialization.
+ *
+ * There is no defensive copy when reading from the database as protobufs are assumed to
+ * be immutable.
```
Comment (swankjesse) [inline misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumn.kt]: 👍🏻

File: misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumnType.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    session: SharedSessionContractImplementor?,
+    owner: Any?
+  ): Any? {
+    val result = rs?.getBytes(names[0])
+    return if (result != null) protoAdapter.decode(result) else null
+  }
+
+  override fun isMutable() = false
+
+  override fun sqlTypes() = intArrayOf(Types.VARCHAR)
```
Comment (swankjesse) [inline misk-hibernate/src/main/kotlin/misk/hibernate/ProtoColumnType.kt]: not VARBINARY ?

**Comment by swankjesse** [comment]
Nice!

**Pull request merged**